src/enzyme_ad/jax/Passes/DialectCanonicalize.cpp

```diff
@@ -0,0 +1,56 @@
+//===- DialectCanonicalize.cpp - Dialect-restricted canonicalizer pass ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "src/enzyme_ad/jax/Passes/Passes.h"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_DIALECTCANONICALIZE
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+namespace {
+struct DialectCanonicalizePass
+    : mlir::enzyme::impl::DialectCanonicalizeBase<DialectCanonicalizePass> {
+  using DialectCanonicalizeBase::DialectCanonicalizeBase;
+
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    for (const std::string &dialectName : dialects) {
+      if (Dialect *dialect = getContext().getLoadedDialect(dialectName)) {
```

Comment:
if there's any {llvm, enzymexla, memref, arith, affine} it may be worthwhile to load the whole set (per some patterns generating code in other dialects, and also enzymexla hijacking to add patterns applying to other dialects iirc)

Reply:
You can list those in pass options. As many as you want, there is no hardcoding here. But you definitely don't need tensor and tosa canonicalizers

Reply:
ohh I read to quickly here, I thought this gathered the options based off of ops in IR, not a pass string, this makes more sense